[c10d] Make send/recv as custom ops

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79779

Summary:
This patch makes send/recv as custom ops such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_send_recv
...and other existing distributed tests.

Commit:
[c10d] Make send/recv as custom ops

Summary:
This patch makes send/recv as custom ops such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_send_recv
...and other existing distributed tests.

[ghstack-poisoned]

Commit:
Update on "[c10d] Make send/recv as custom ops"


Summary:
This patch makes send/recv as custom ops such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_send_recv
...and other existing distributed tests.

[ghstack-poisoned]

Commit:
Update on "[c10d] Make send/recv as custom ops"


Summary:
This patch makes send/recv as custom ops such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_send_recv
...and other existing distributed tests.

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -118,6 +118,26 @@ c10::intrusive_ptr<ProcessGroup::Work> barrier(
 }
 
+c10::intrusive_ptr<ProcessGroup::Work> send(
+    at::TensorList tensors,
```
same comment on `TensorList` vs `std::vector<Tensor>`

alanwaketan replied:
I think it's a convention to use TensorList to represent const std::vectorat::Tensor&.

I only use const std::vectorat::Tensor& if there is a const std::vector<std::vectorat::Tensor>& in the signature to keep consistency on that function. Let me know which way you like.

mrshenli commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -118,6 +118,26 @@ c10::intrusive_ptr<ProcessGroup::Work> barrier(
       BarrierOptions{device_ids, std::chrono::milliseconds(timeout)});
 }
 
+c10::intrusive_ptr<ProcessGroup::Work> send(
```
trailing underscore?

alanwaketan replied:
Send is not a in-place op. So there is no needs to have a trailing _.

wanchaol commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -328,5 +350,35 @@ c10::intrusive_ptr<ProcessGroup::Work> barrier(
+    int64_t dstRank,
+    int64_t tag) {
+  static auto op = c10::Dispatcher::singleton()
+                       .findSchemaOrThrow("c10d::send", "")
```
hmm would we not find the schema by any chance?

alanwaketan replied:
Why would you have such concern? Do you happen to know another op called "c10d::send" as well?

wanchaol commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -328,5 +350,35 @@ c10::intrusive_ptr<ProcessGroup::Work> barrier(
+  return op.call(tensors, process_group, dstRank, tag);
+}
+
+c10::intrusive_ptr<ProcessGroup::Work> recv(
```
just noticed we have this `recv` and `recv_` two separate APIs (above), what's the difference between these two and should we merge?

alanwaketan replied:
The above recv_ is the implementation of the function schema. This one is the C++ API for users to call the op. recv_ follows aten ops's conventions while recv follows c10d's conventions.

wanchaol reviewed (approved):
lgtm, have some questions.

Commit:
Update on "[c10d] Make send/recv as custom ops"


Summary:
This patch makes send/recv as custom ops such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_send_recv
...and other existing distributed tests.

[ghstack-poisoned]

alanwaketan commented:
@pytorchbot merge --green

alanwaketan commented:
Thanks for approving this pull request, Shen and Wanchao.

Commit:
Update on "[c10d] Make send/recv as custom ops"


Summary:
This patch makes send/recv as custom ops such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_send_recv
...and other existing distributed tests.

[ghstack-poisoned]

alanwaketan commented:
@pytorchbot merge --green
